crates/doc/src/inference.rs

```diff
@@ -1484,6 +1486,201 @@ impl Shape {
+                            accum || addl.widen_inner(node.value(), loc.push_item(idx))
+                        })
+                    }
+                    _ => {
+                        let (hint, new_obj_shape) = itertools::merge_join_by(
+                            self.object.properties.iter_mut(),
+                            obj.iter(),
+                            |prop, field| prop.name.cmp(&field.property().to_string()),
+                        )
+                        .fold(
+                            // NOTE: We need to do this whole song and dance of folding over `ObjShape` because
+                            // in `EitherOrBoth::Right`, we don't have a `lhs` to mutate, and we can't directly mutate
+                            // `self.object` because we're iterating over it, and `.fold` requires unique access.
+                            // This gives us a temporary blank ObjShape into which we can add all brand new fields
+                            // which will then get union()'d into `self.object` after we're done folding.
+                            (false, {
+                                let mut blank_shape = ObjShape::new();
+                                blank_shape.additional = Some(Box::new(Shape {
+                                    type_: types::INVALID,
+                                    provenance: Provenance::Inline,
+                                    ..Default::default()
+                                }));
+                                blank_shape
+                            }),
+                            |(accum, new_obj_shape), eob| match eob {
+                                // Both the shape and node have this field, recursion time
+                                EitherOrBoth::Both(lhs, rhs) => (
+                                    accum
+                                        || lhs.shape.widen_inner(
+                                            rhs.value(),
+                                            loc.push_prop(rhs.property()),
+                                        ),
+                                    new_obj_shape,
+                                ),
+                                // Shape has a field that the node is missing, so let's make sure it's not marked as required
+                                // Is this the right behavior here?
+                                EitherOrBoth::Left(mut lhs) => {
+                                    lhs.is_required = false;
+                                    (accum, new_obj_shape)
+                                }
+                                // The Node has a field that the shape doesn't, let's add it
+                                EitherOrBoth::Right(rhs) => {
+                                    let mut prop = ObjProperty {
+                                        name: rhs.property().to_owned(),
+                                        // Should we really do this?
```

Comment:
no, it's not required with one exception, which is when we're merging in the very first instance, which needs to be accounted for. Also do you have a test on this case?

Reply:
We talked, I understand what this means now: by definition, only properties that we have encountered on _every_ object should be marked as required. Therefore, only properties we encounter the _very first time_ we see an object at a particular location should get `is_required: true`, since any other time the implication is that we have not seen this property before, i.e we've seen documents without it, and therefore it should not be marked as required.